tests/agent_server/test_new_event_mechanism.py

```diff
@@ -0,0 +1,624 @@
+    The dependant resource should not be redeployed
+    """
+    model, version = initial_deployment
+
+    # EventClient setup
+    event_client = EventClient(client, environment)
+
+    # Recreating same model, with incremented version
+    version = await clienthelper.get_version()
+    model = init_model(version)
+    resources = list(model.values())
+
+    # Exporting new version
+    response: Result = await client.put_version(
+        tid=environment,
+        version=version,
+        resources=resources,
+        unknowns=[],
+        version_info={},
+        compiler_version=get_compiler_version(),
+    )
+    assert response.code == 200
+
+    # Deploying
+    response: Result = await client.release_version(
+        environment,
+        version,
+        True,
+        const.AgentTriggerMethod.push_full_deploy,
+    )
+    assert response.code == 200
+    await _wait_until_deployment_finishes(client, environment, version)
+
+    resource_id = Id.parse_id(model["root"]["id"])
+    last_change = await event_client.get_resource_action(resource_id, is_deployment_with_change, oldest_first=False)
+    first_change = await event_client.get_resource_action(resource_id, is_deployment_with_change, oldest_first=True)
+
+    # We check that we didn't have a second deployment
+    assert last_change == first_change
+    assert last_change.change == Change.updated
+    assert last_change.version == version - 1
+
+
+@pytest.mark.asyncio
+async def test_repair(
+    resource_container: ResourceContainer,
+    environment: UUID,
+    server,
+    client: Client,
+    agent,
+    clienthelper: ClientHelper,
+    initial_deployment: Tuple[dict, int],
+    no_agent_backoff,
+):
+    """
+    Testing a repair, doing a full deployment of an existing version.
+    Only one dependency value is "altered" (to be fixed by the repair)
+    The dependant resource should be redeployed.
+    """
+    model, version = initial_deployment
+
+    # EventClient setup
+    event_client = EventClient(client, environment)
+
+    # Manually changing resource values
+    resource_id = Id.parse_id(model["dep1"]["id"])
+    resource_container.Provider.set(resource_id.agent_name, resource_id.attribute_value, "updated")
+
+    response = await client.get_version(environment, version, True)
+    assert response.code == 200
+    last_deployment_date = sorted([res["last_deploy"] for res in response.result["resources"]])[-1]
+
+    # Triggering repair
+    response: Result = await client.deploy(
+        environment,
+        const.AgentTriggerMethod.push_full_deploy,
+    )
+    assert response.code == 200
+
+    async def is_repair_finished():
+        response = await client.get_version(environment, version)
+
+        if "deploying" in [res["status"] for res in response.result["model"]["status"].values()]:
+            return False
+
+        for res in response.result["resources"]:
+            if res["last_deploy"] > last_deployment_date:
+                return True
```

Comment:
this means that the repair is done as soon as one resource got deployed?

Reply:
And if there is not resource marked as "deploying" (see the previous condition)

Reply:
that is sensitive to race conditions (there is a short time between a resource being done and the next going to deploying)

beter to check here that all last_deploy > last_deployment_dat

Reply:
Done in https://github.com/inmanta/inmanta-core/pull/2762/commits/f97a03e9ff406aa88ba41c2a9d3db28c8b68128c